pkg/media/whip/whipsrc.go

```diff
@@ -0,0 +1,371 @@
+		if err := webrtcSettings.SetEphemeralUDPPortRange(icePortStart, icePortEnd); err != nil {
+			return nil, err
+		}
+	}
+	webrtcSettings.SetIncludeLoopbackCandidate(p.Whip.EnableLoopbackCandidate)
+
+	m, err := newMediaEngine(p)
+	if err != nil {
+		return nil, err
+	}
+
+	// Create a InterceptorRegistry. This is the user configurable RTP/RTCP Pipeline.
+	// This provides NACKs, RTCP Reports and other features. If you use `webrtc.NewPeerConnection`
+	// this is enabled by default. If you are manually managing You MUST create a InterceptorRegistry
+	// for each PeerConnection.
+	i := &interceptor.Registry{}
+
+	// Use the default set of Interceptors
+	if err := webrtc.RegisterDefaultInterceptors(m, i); err != nil {
+		return nil, err
+	}
+
+	// Create the API object with the MediaEngine
+	api := webrtc.NewAPI(webrtc.WithMediaEngine(m), webrtc.WithSettingEngine(*webrtcSettings), webrtc.WithInterceptorRegistry(i))
+	s.pc, err = s.createPeerConnection(api)
+	if err != nil {
+		return nil, err
+	}
+
+	sdpAnswer, err := s.getSDPAnswer(ctx, offer)
+	if err != nil {
+		return nil, err
+	}
+
+	err = s.postSDPAnswer(ctx, sdpAnswer)
+	if err != nil {
+		return nil, err
+	}
+
+	return s, nil
+}
+
+func (s *WHIPSource) Start(ctx context.Context) error {
+	s.trackLock.Lock()
+	defer s.trackLock.Unlock()
+
+	for _, v := range s.trackSrc {
+		err := v.Start()
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (s *WHIPSource) Close() error {
+	if s.pc == nil {
+		return nil
+	}
+
+	s.pc.Close()
+
+	return <-s.result
+}
+
+func (s *WHIPSource) GetSources(ctx context.Context) []*app.Source {
+	for {
+		s.trackLock.Lock()
+		if len(s.trackSrc) >= s.expectedTrackCount {
+			ret := make([]*app.Source, 0)
+
+			for _, v := range s.trackSrc {
+				ret = append(ret, v.GetAppSource())
+			}
+
+			s.trackLock.Unlock()
+			return ret
+		}
+		s.trackLock.Unlock()
+
+		select {
+		case <-ctx.Done():
+			logger.Infow("GetSources cancelled before all sources were ready")
+			return nil
+		case <-s.trackAddedChan:
+			// continue
+		}
+	}
+}
+
+func (s *WHIPSource) createPeerConnection(api *webrtc.API) (*webrtc.PeerConnection, error) {
+	config := webrtc.Configuration{
+		ICEServers:   []webrtc.ICEServer{},
+		SDPSemantics: webrtc.SDPSemanticsUnifiedPlanWithFallback,
+		BundlePolicy: webrtc.BundlePolicyBalanced,
+	}
+
+	// Create a new RTCPeerConnection
+	pc, err := api.NewPeerConnection(config)
+	if err != nil {
+		return nil, err
+	}
+
+	// Accept one audio and one video track incoming
+	for _, typ := range []webrtc.RTPCodecType{webrtc.RTPCodecTypeVideo, webrtc.RTPCodecTypeAudio} {
+		if _, err := pc.AddTransceiverFromKind(typ, webrtc.RTPTransceiverInit{
+			Direction: webrtc.RTPTransceiverDirectionRecvonly,
+		}); err != nil {
+			return nil, err
+		}
+	}
+
+	pc.OnTrack(s.addTrack)
+
+	closeOnce := sync.Once{}
+	pc.OnConnectionStateChange(func(state webrtc.PeerConnectionState) {
+		logger.Infow("Peer Connection State changed", "state", state.String())
```

Comment:
nit: would be better to use whipsrc logger with the right fields for the current ingress.